src/libraries/System.Runtime.Numerics/src/System/Number.BigInteger.cs

```diff
@@ -783,44 +780,22 @@ internal static bool TryFormatBigInteger(BigInteger value, ReadOnlySpan<char> fo
                 }
             }
 
-            // First convert to base 10^9.
-            int cuSrc = value._bits.Length;
-            // A quick conservative max length of base 10^9 representation
-            // A uint contributes to no more than 10/9 of 10^9 block, +1 for ceiling of division
-            int cuMax = cuSrc * (MaxPartialDigits + 1) / MaxPartialDigits + 1;
-            Debug.Assert((long)BigInteger.MaxLength * (MaxPartialDigits + 1) / MaxPartialDigits + 1 < (long)int.MaxValue); // won't overflow
-
-            uint[]? bufferToReturn = null;
-            Span<uint> base1E9Buffer = cuMax < BigIntegerCalculator.StackAllocThreshold ?
-                stackalloc uint[cuMax] :
-                (bufferToReturn = ArrayPool<uint>.Shared.Rent(cuMax));
-
-            int cuDst = 0;
+            // The Ratio is calculated as: log_{10^9}(2^32)
+            const double digitRatio = 1.0703288734719332;
```

Comment:
Does it worth to do a FP calculation, instead of conservatively allocate for upper limit?

Reply:
There's not much difference, but the required storage size will be smaller. For example, when the bit length is 58 or 59, stackalloc is used.

|_bit.Length|old|new|
|---|---|---|
|57|64|62|
|58|65|63|
|59|66|64|

This change also aims to achieve mathematical correctness in the implementation.

Reply:
Either stack allocation or array poll will round the requested size up to whole number. Such difference should be really negligible.

Reply:
I rounded up the value of `digitRatio` from $1.0703288734719332$, which is the return value of `Math.Log(1L<<32, 1e9)`, to $1.070328873472$. Since `double` has higher precision than `int`, unintended rounding down has become theoretically impossible.